### DISCUSSION TITLE
How to integrate `InstrumentChannel` class into `VisaInstrument` class

### DISCUSSION THREAD

**Discussion by paniash** [microsoft/Qcodes#5595] [Q&A] [answered]
Hi all,

I'm very new to QCoDeS, so kindly excuse my ignorance.

I'm writing a driver for an oscilloscope with four channels and have defined an `InstrumentChannel` class containing all the parameters and another `VisaInstrument` class following the [example](https://qcodes.github.io/Qcodes/examples/writing_drivers/Creating-Instrument-Drivers.html#VisaInstrument:-a-more-involved-example) given in the docs.

However, when writing a script by importing the `VisaInstrument` class, I'm not exactly sure how I can define channels in the main `VisaInstrument` class. 

Here's the relevant driver file (`keysight/Keysight_DSOX4104A.py` and `test-script.py`): https://gitlab.com/paniash/qcode-drivers/-/tree/main/keysight

Thanks in advance!

**Comment by jenshnielsen** [comment]
If you look at the example that you have linked you should see inside
the `Keithley2600` class.

```
    # Add the channel to the instrument
    for ch in ["a", "b"]:
        ch_name = f"smu{ch}"
        channel = KeithleyChannel(self, ch_name, ch_name)
        self.add_submodule(ch_name, channel)
```

This addes submodules `smua` and `smub` that you can then access as instrument.a and instrument.b when you have initialized the instrument.

**Reply by paniash** [reply]
Hi @jenshnielsen! Thanks for the reply. I'm trying to define the instrument with
```python
osc = KeysightDSOX4104A(name="osc", address="TCPIP0::a-dx4104a-60137::INSTR",
                        channel=[1, 2, 3, 4])
```

However, I get an error stating that a `channel` attribute does not exist. How would I go ahead defining channels in this case?

**Reply by jenshnielsen** [reply, marked as answer]
In the code that you have linked you have 


```
class KeysightDSOX4104A(VisaInstrument):
    """
    This is the qcodes driver for the Keysight InfiniiVision DSOX4104A oscilloscope
    """

    def __init__(self, name: str, address: str, **kwargs) -> None:
        """
        Args:
            name: Name used internally by QCoDeS
            address: VISA resource address
        """
        super().__init__(name, address, terminator="\n", **kwargs)

        # Add channel to instrument
        for ch in [1, 2, 3, 4]:
            ch_name = f"ch{ch}"
            channel = DSOX4104AChannel(self, ch_name, ch)
            self.add_submodule(ch, channel)
            self.channels.append(channel)


```

Here you are trying to add the channel that you have created to a list of channels in the last line but you have never devined than list any where. 

Now there are two alternatives. You don't need that line without it you will still be able to access the channels as osc.ch1 etc.

You only need that line if you want to be able to access the channels by index or set a value on more than one channel in the same operation. E.g do `osc.channels[0]` to do that you should make that variable a ChannelList or ChannelTuple. 

https://qcodes.github.io/Qcodes/api/instrument/index.html#qcodes.instrument.ChannelTuple

One way could be:


```
class KeysightDSOX4104A(VisaInstrument):
    """
    This is the qcodes driver for the Keysight InfiniiVision DSOX4104A oscilloscope
    """

    def __init__(self, name: str, address: str, **kwargs) -> None:
        """
        Args:
            name: Name used internally by QCoDeS
            address: VISA resource address
        """
        super().__init__(name, address, terminator="\n", **kwargs)
        channels = []
        # Add channel to instrument
        for ch in [1, 2, 3, 4]:
            ch_name = f"ch{ch}"
            channel = DSOX4104AChannel(self, ch_name, ch)
            self.add_submodule(ch, channel)
            channels.append(channel)
       channel_tuple = ChannelTuple(self, "channels", DSOX4104AChannel, channels)
       self.add_submodule("channels", channel_tuple)

```


You do not need to and should not pass `channel=[1, 2, 3, 4]` as an argument when you initialize the Instrument Class

**Reply by paniash** [reply]
Thanks! May I ask about the other alternative where I could access the channels as `osc.ch1`? Currently without your suggested code, I get the error:
```
'KeysightDSOX4104A' object and its delegates have no attribute 'ch1'
```

**Reply by jenshnielsen** [reply]
Could you check the value of `osc.submodules` this should be a dictionary that maps from the name `ch1` to the modules/channels 
if things work correctly.

**Reply by paniash** [reply]
Ah, I see. So the output of `osc.submodules` yields:
```
{1: <DSOX4104AChannel: osc_ch1 of KeysightDSOX4104A: osc>, 2: <DSOX4104AChannel: osc_ch2 of KeysightDSOX4104A: osc>, 3: <DSOX4104AChannel: osc_ch3 of KeysightDSOX4104A: osc>, 4: <DSOX4104AChannel: osc_ch4 of KeysightDSOX4104A: osc>, 'channels': ChannelTuple(<KeysightDSOX4104A: osc>, DSOX4104AChannel, [<DSOX4104AChannel: osc_ch1 of KeysightDSOX4104A: osc>, <DSOX4104AChannel: osc_ch2 of KeysightDSOX4104A: osc>, <DSOX4104AChannel: osc_ch3 of KeysightDSOX4104A: osc>, <DSOX4104AChannel: osc_ch4 of KeysightDSOX4104A: osc>])}
```

My channels are named 1, 2, 3, 4 instead of ch1, ch2, ch3, ch4. Makes sense why I can't access them as `osc.1`.

**Reply by jenshnielsen** [reply]
That is caused by your line 180. Here `self.add_submodule(ch, channel)` should have been `self.add_submodule(ch_name, channel)` 

If you do that you can remove these lines


```python
        self.ch1 = self.channels[0]
        self.ch2 = self.channels[1]
        self.ch3 = self.channels[2]
        self.ch4 = self.channels[3]
```
I recommend doing that since that will ensure that your snapshot is consistent

**Reply by paniash** [reply]
Great! Thanks a lot @jenshnielsen. :)
